kernel/src/mem.rs

```diff
@@ -69,213 +82,199 @@ unsafe fn raw_appslice_to_slice<'a>(ptr: *const u8, len: usize) -> &'a [u8] {
+    ///
+    /// A default instance of a process buffer must return a pointer
+    /// to address `0x0`.
     fn ptr(&self) -> *const u8;
 
-    /// Applies a function to the (read only) slice reference pointed
-    /// to by the AppSlice.
+    /// Applies a function to the (read only) process slice reference
+    /// pointed to by the process buffer.
     ///
     /// If the process is no longer alive and the memory has been
     /// reclaimed, this method must return the default value.
     ///
-    /// # Default AppSlice
+    /// # Default Process Buffer
     ///
-    /// A default instance of an AppSlice must return the passed
-    /// default value without executing the closure.
+    /// A default instance of a process buffer must return the passed
+    /// default value without executing the passed closure.
     fn map_or<F, R>(&self, default: R, fun: F) -> R
     where
-        F: FnOnce(&[u8]) -> R;
+        F: FnOnce(&ReadableProcessSlice) -> R;
 }
 
-/// A readable and writable region of userspace memory.
+/// A readable and writeable region of userspace process memory.
 ///
 /// This trait can be used to gain read-write access to memory regions
-/// wrapped in an AppSlice type.
+/// wrapped in a [`ReadWriteProcessBuf`].
 ///
-/// This is a supertrait of [`Read`], which features further methods.
-pub trait ReadWrite: Read {
-    /// Applies a function to the mutable slice reference pointed to
-    /// by the AppSlice.
+/// This is a supertrait of [`ReadableProcessBuf`], which features
+/// methods allowing mutable access.
+pub trait WriteableProcessBuf: ReadableProcessBuf {
+    /// Applies a function to the mutable process slice reference
+    /// pointed to by the [`ReadWriteProcessBuf`].
     ///
     /// If the process is no longer alive and the memory has been
     /// reclaimed, this method must return the default value.
     ///
-    /// # Default AppSlice
+    /// # Default Process Buffer
     ///
-    /// A default instance of an AppSlice must return the passed
+    /// A default instance of a process buffer must return the passed
     /// default value without executing the closure.
-    fn mut_map_or<F, R>(&mut self, default: R, fun: F) -> R
+    fn mut_map_or<F, R>(&self, default: R, fun: F) -> R
     where
-        F: FnOnce(&mut [u8]) -> R;
+        F: FnOnce(&WriteableProcessSlice) -> R;
 }
 
-/// Read-writable memory region of a process, shared with the kernel
-pub struct ReadWriteAppSlice {
-    ptr: *mut u8,
+/// Read-only buffer shared by a userspace process
+///
+/// This struct is provided to capsules when a process Allows a
+/// particular section of its memory to the kernel and gives the
+/// kernel read access to this memory.
+///
+/// It can be used to obtain a [`ReadableProcessSlice`], which is
+/// based around a slice of [`Cell`]s. This is because a userspace can
+/// Allow overlapping sections of memory into different
+/// [`ReadableProcessSlice`]. Having at least one mutable Rust slice
```

Comment:
```suggestion
/// allow overlapping sections of memory into different
/// [`ReadableProcessSlice`]s. Having at least one mutable Rust slice
```

Reply:
I have taken care to (up to now) capitalize _allow_ explicitly, as I wanted to emphasize I was talking about Tock's mechanism and now "allow" aka. "permit". Is that even a correct thing to do in English? Happy to change it.

Reply:
Thinking about it, it probably makes more sense to distinguish those through a Markdown inline-code syntax: "`allow` overlapping"